addons/godot_rl_agents/sync.gd

```diff
@@ -2,165 +2,253 @@ extends Node
+						"has no onnx model path set.",
+						"Using path set on the sync node instead."
+				)
+				agent_onnx_model = onnx_models[onnx_model_path]
+			else:
+				if not onnx_models.has(agent.onnx_model_path):
+					assert(
+						FileAccess.file_exists(agent.onnx_model_path),
+						(
+							"Onnx Model Path set on %s node does not exist: %s"
+							% [agent.get_path(), agent.onnx_model_path]
+						)
+					)
+					onnx_models[agent.onnx_model_path] = ONNXModel.new(agent.onnx_model_path, 1)
+				agent_onnx_model = onnx_models[agent.onnx_model_path]
+
+			agent.onnx_model = agent_onnx_model
+		_set_heuristic("model", agents_inference)
+
+func _physics_process(_delta):
 	# two modes, human control, agent control
 	# pause tree, send obs, get actions, set actions, unpause tree
 	if n_action_steps % action_repeat != 0:
 		n_action_steps += 1
 		return
 
 	n_action_steps += 1
-	
+
+	_training_process()
+	_inference_process()
+
+
+func _training_process():
 	if connected:
-		get_tree().set_pause(true) 
-		
+		get_tree().set_pause(true)
+
 		if just_reset:
 			just_reset = false
-			var obs = _get_obs_from_agents()
-		
-			var reply = {
-				"type": "reset",
-				"obs": obs
-			}
+			var obs = _get_obs_from_agents(agents_training)
+
+			var reply = {"type": "reset", "obs": obs}
 			_send_dict_as_json_message(reply)
 			# this should go straight to getting the action and setting it checked the agent, no need to perform one phyics tick
-			get_tree().set_pause(false) 
+			get_tree().set_pause(false)
 			return
-		
+
 		if need_to_send_obs:
 			need_to_send_obs = false
 			var reward = _get_reward_from_agents()
 			var done = _get_done_from_agents()
 			#_reset_agents_if_done() # this ensures the new observation is from the next env instance : NEEDS REFACTOR
-			
-			var obs = _get_obs_from_agents()
-			
-			var reply = {
-				"type": "step",
-				"obs": obs,
-				"reward": reward,
-				"done": done
-			}
+
+			var obs = _get_obs_from_agents(agents_training)
+
+			var reply = {"type": "step", "obs": obs, "reward": reward, "done": done}
 			_send_dict_as_json_message(reply)
-		
+
 		var handled = handle_message()
-	
-	elif onnx_model != null:
-		var obs : Array = _get_obs_from_agents()
-	
+
+
+func _inference_process():
+	if agents_inference.size() > 0:
+		var obs: Array = _get_obs_from_agents(agents_inference)
 		var actions = []
-		for o in obs:
-			var action = onnx_model.run_inference(o["obs"], 1.0)
+
+		for agent_id in range(0, agents_inference.size()):
+			var action = agents_inference[agent_id].onnx_model.run_inference(
+				obs[agent_id]["obs"], 1.0
+			)
 			action["output"] = clamp_array(action["output"], -1.0, 1.0)
-			var action_dict = _extract_action_dict(action["output"])
+			var action_dict = _extract_action_dict(
+				action["output"], _action_space_inference[agent_id]
+			)
 			actions.append(action_dict)
-		
-		_set_agent_actions(actions) 
+
+		_set_agent_actions(actions, agents_inference)
 		need_to_send_obs = true
```

Comment:
Does the `need_to_send_obs` flag still need to be set in the `_inference_process` method?

Reply:
I think I fixed this later in the IL branch. I'll have to check and bring any fixes from there.